src/qutip_tensornetwork/core/data/network.py

```diff
@@ -0,0 +1,627 @@
+            in_edges and out_edges.
+
+        copy: bool, default True
+            Whether to copy all the ``Nodes``/``Edges`` involved in the
+            network.
+        """
+        if (len(in_edges) == 0 and len(out_edges) == 0
+            and (nodes is None or len(nodes)==0)):
+            raise ValueError("Since no edges were provided, it was not possible"
+                             "to infer which nodes belong to the network."
+                             "You may want to include a scalar node to represent"
+                             "a matrix with shape (1,1).")
+
+        self.out_edges = list(out_edges)
+        self.in_edges = list(in_edges)
+
+        # Unlike in QuOperator we will keep track of the nodes instead of
+        # dynamically searching for them when necessary. This is because
+        # searching all nodes in a large graph can be quite expensive while
+        # keeping track of them with network operations is straightforward.
+        self.nodes = set(nodes) if nodes else tn.reachable(self.in_edges
+                                                         + self.out_edges)
+
+        self._check_edge_nodes_in_nodes()
+        self._check_in_out_are_dangling()
+        self._check_only_in_out_are_dangling()
+        self._check_edges_unique()
+        # I may need extra check such as for backends etc in the future
+
+        if copy:
+            node_dict, edge_dict = tn.copy(self.nodes)
+            self.nodes = set(node_dict[n] for n in self.nodes)
+            self.in_edges = [edge_dict[e] for e in self.in_edges]
+            self.out_edges = [edge_dict[e] for e in self.out_edges]
+
+    @property
+    def shape(self):
+        return (np.prod(self.dims[0], dtype=int), np.prod(self.dims[1],
+                                                          dtype=int))
+
+    @property
+    def dims(self):
+        out_space = [e.dimension for e in self.out_edges]
+        in_space = [e.dimension for e in self.in_edges]
+        return [out_space, in_space]
+
+    def _check_in_out_are_dangling(self):
+        for (i, e) in enumerate(self.out_edges):
+          if not e.is_dangling():
+            raise ValueError("output edge {} is not dangling!".format(i))
+        for (i, e) in enumerate(self.in_edges):
+          if not e.is_dangling():
+            raise ValueError("input edge {} is not dangling!".format(i))
+
+    def _check_only_in_out_are_dangling(self):
+        known_edges = set(self.in_edges + self.out_edges)
+        all_dangling_edges = get_subgraph_dangling(self.nodes)
+        if known_edges != all_dangling_edges:
+            unexpected_edges = all_dangling_edges.difference(known_edges)
+            raise ValueError("the network includes unexpected dangling edges."
+                            + str(unexpected_edges))
+
+    def _check_edges_unique(self):
+        """Check that in_edges and out_edges are unique."""
+        if (len(set(self.in_edges)) != len(self.in_edges) or
```

Comment:
Should `in_edges` and `out_edges` not be unique when combined and not just separately? I.e should `len(set(in_edges + out_edges)) == len(in_edges) + len(out_edges)`?

Reply:
Indeed, good catch. We do not want someone including the same edge to both in_edges and out_edges